pkg/kubectl/cmd/util/config_flags.go

```diff
@@ -0,0 +1,363 @@
+	}
+
+	if f.RequireMatchedServerVersion != nil {
+		flags.BoolVar(f.RequireMatchedServerVersion, FlagMatchBinaryVersion, *f.RequireMatchedServerVersion, "Require server version to match client version")
+	}
+}
+
+func (f *ConfigFlags) WithDeprecatedPasswordFlag() *ConfigFlags {
+	f.Username = stringptr("")
+	f.Password = stringptr("")
+	return f
+}
+
+func NewConfigFlags() *ConfigFlags {
+	impersonateGroup := []string{}
+	insecure := false
+	requireMatchServerVersion := false
+
+	return &ConfigFlags{
+		Insecure:   &insecure,
+		Timeout:    stringptr("0"),
+		KubeConfig: stringptr(""),
+
+		ClusterName:                 stringptr(""),
+		AuthInfoName:                stringptr(""),
+		Context:                     stringptr(""),
+		Namespace:                   stringptr(""),
+		APIServer:                   stringptr(""),
+		CertFile:                    stringptr(""),
+		KeyFile:                     stringptr(""),
+		CAFile:                      stringptr(""),
+		BearerToken:                 stringptr(""),
+		Impersonate:                 stringptr(""),
+		RequireMatchedServerVersion: &requireMatchServerVersion,
+		ImpersonateGroup:            &impersonateGroup,
+	}
+}
+
+func stringptr(val string) *string {
+	return &val
+}
+
+type TestConfigFlags struct {
+	loadingRules *clientcmd.ClientConfigLoadingRules
+	clientConfig clientcmd.ClientConfig
+}
+
+func (f *TestConfigFlags) ToRawKubeConfigLoader() clientcmd.ClientConfig {
+	if f.clientConfig != nil {
+		return f.clientConfig
+	}
+
+	loadingRules := f.loadingRules
+	if loadingRules == nil {
+		loadingRules = &clientcmd.ClientConfigLoadingRules{
+			MigrationRules: map[string]string{},
+		}
+	}
+
+	overrides := &clientcmd.ConfigOverrides{ClusterDefaults: clientcmdapi.Cluster{Server: "http://localhost:8080"}}
+	fallbackReader := bytes.NewBuffer([]byte{})
+
+	clientConfig := clientcmd.NewInteractiveDeferredLoadingClientConfig(loadingRules, overrides, fallbackReader)
+	return clientConfig
+}
+
+func (f *TestConfigFlags) ToRESTConfig() (*rest.Config, error) {
+	return f.ToRawKubeConfigLoader().ClientConfig()
+}
+
+func (f *TestConfigFlags) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
+	config, err := f.ToRESTConfig()
+	if err != nil {
+		return nil, err
+	}
+	config.Burst = 100
+
+	cacheDir := filepath.Join(homedir.HomeDir(), ".kube", "http-cache")
+	wt := config.WrapTransport
+	config.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
+		if wt != nil {
+			rt = wt(rt)
+		}
+		return transport.NewCacheRoundTripper(cacheDir, rt)
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	cacheDir = computeDiscoverCacheDir(filepath.Join(homedir.HomeDir(), ".kube", "cache", "discovery"), config.Host)
+	return NewCachedDiscoveryClient(discoveryClient, cacheDir, time.Duration(10*time.Minute)), nil
```

Comment:
memcache discovery at most

Reply:
@deads2k
> memcache discovery at most

not sure what you mean by this

Reply:
For a fake test client, don't write to disk. Just use the memory caching client

Reply:
ah, wrong spot in the file.  weird.